modules/system/database/migrations/2021_10_02_000026_Db_Media_Items.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+use Winter\Storm\Database\Schema\Blueprint;
+use Winter\Storm\Database\Updates\Migration;
+
+class DbMediaItems extends Migration
+{
+    public function up()
+    {
+        Schema::create('media_items', function (Blueprint $table) {
+            // Schema
+            $table->engine = 'InnoDB';
+            $table->increments('id');
+            $table->integer('parent_id')->unsigned()->nullable();
+            $table->integer('nest_left')->unsigned()->nullable();
+            $table->integer('nest_right')->unsigned()->nullable();
+            $table->integer('nest_depth')->unsigned()->nullable();
```

Comment:
I'm concerned keeping the nested tree data in sync with the path would be more complicated than just using the path by itself. We should be able to do queries based on the path, and we wouldn't want order inside of a path to be determined by integers in the database rather than the file attributes like path, size, date modified, etc.

Reply:
I felt using nested set will make it easy to modify hierarchical data in bulk, as opposed to using any form of `parent_id` or trying to substring a path to work out multiple levels of child structure.

I definitely wasn't going to use it for ordering, that'll just be name, size, date modified as it is now.

Reply:
Well I'll wait to see the implementation but I'm concerned about the reliability

Reply:
A competing CMS solved this using a variation of path enumeration:
https://github.com/drupal/drupal/blob/9.3.x/core/lib/Drupal/Core/Menu/MenuTreeStorage.php#L529 (they do limit the depth to 10 levels)

Also based on the below comparison nested sets are somewhat slower
![image](https://user-images.githubusercontent.com/966581/135609001-210f7c6b-f04c-45e7-a576-9392c3b949e9.png)

Reply:
@lex0r really sorry for the delay in responding to this.

Thanks for posting that, however, I slightly disagree with what it puts forward. While nested sets might be hard to wrangle if you build the functionality from scratch, we already have a [NestedTree trait](https://github.com/wintercms/storm/blob/develop/src/Database/Traits/NestedTree.php) which does all the heavy lifting, so it's trivial for us to use.

All the actions - querying, deleting, inserting and moving - are simple through this API.

I'm not quite sure what the "referential integrity" column refers to, however, my previous experience with nested trees finds them to be incredibly reliable and auditable. They can even be easily repaired.

The SimpleTree trait handles the "Adjacency List" design you mention, but as your diagram mentions, it's harder to query a subtree, and more to the point, that's a design fault - you have to run several queries in order to find the full subtree, which is not useful in this case if we want to do filtering and searching. Nested set, by design, handles this flawlessly.

Reply:
Hi @bennothommo 
I agree with your argument that NestedTree trait is trivial to use, and I understand your experience using them may be positive, but I'd like to share my experience with them in a real world project. To keep it short, nested sets don't survive parallel modifications of the tree structure, and if that happens they are guaranteed to get corrupted. For me it's a big no when it comes to applications used by more than 1 person. I appreciate they are easy to fix (thanks to the parent ref) but why fixing what shouldn't even break?

On the other hand, path enumeration (also known as materialised path) is used in many CMS like Drupal (sorry to repeating that again) and it handles parallel modifications to the trees easily. To make it more attractable, I will mention that I was able to easily implement it using a very slightly modified version of https://github.com/vicklr/materialized-model.

Hope you could give it a try.

Reply:
@bennothommo benchmarking should help, and an obvious performance improvement for the path enumeration concept is to have all path components as individual columns, eg.  p1, p2, p3 .. p10, where pN is an ID rather than a string. This will limit the depth but rarely people need a super deep nesting. Alternatively, having an index on the path is a scalable approach especially when talking about the ID-based path (/1/123/544) not the file path  (/content/images/abc). The index will be small as only IDs are stored, which are 4 bytes max, unlike the strings, and it will be used in all searches starting with a known prefix (which I guess is always known).

Table locking is a big performance killer, I tried that and it left me surprised with how slow it can be, so can't recommend it as a solution for a modern multi-user CMS.